### PR TITLE
Pellets from spread projectiles don't freeze clients

### PR DESCRIPTION
**Pull request by carlarctg**
## About The Pull Request
to be more precise they no longer send 12 messages in a row to your chat that freeze the game via a shitty var that prevents visible_message from appearing

## Why It's Good For The Game
i'm tired of dying not to the bullet, but to the massive freezing buckshot causes
very often it feels like i could have done something if it wasn't for the effective tasing shotguns do when they hit you via pellets
i HIGHLY doubt this is going to make shotguns a 'stealth killer' or anything as they still have 
1. the fuckhuge sprite ninja made
2. the obvious sound they make
3. the fact that combat shotguns are OP and riot shotgun pumps are a joke, making it highly unlikely for an unarmored person to react or defend themselves properly due to the blitzing fast way you get damage from buckshot anyways

## Changelog
:cl:
fix: buckshot no longer freezes your game, but also doesn't show up in chat anymore
/:cl:

### CODE REVIEW

File: code/modules/projectiles/projectile.dm [hunk 1]
```diff
@@ -191,8 +192,9 @@
 			if(hitsound)
 				var/volume = vol_by_damage()
 				playsound(loc, hitsound, volume, TRUE, -1)
-			L.visible_message("<span class='danger'>[L] is hit by \a [src][organ_hit_text]!</span>", \
-					"<span class='userdanger'>You're hit by \a [src][organ_hit_text]!</span>", null, COMBAT_MESSAGE_RANGE)
+			if(lagshot == FALSE)
```
Comment (Couls) [inline code/modules/projectiles/projectile.dm]:
```suggestion
			if(!lagshot)
```

**Comment by SkeletalElite** [comment]
I think getting rid of the buckshot lag is way more important than having a message when you get blasted with shotgun pellets.

**Comment by carlarctg** [comment]
> I think getting rid of the buckshot lag is way more important than having a message when you get blasted with shotgun pellets.

that's what causes the lag, announce message is horribly optimized

**Comment by Kaffe-work** [comment]
Good change

**Comment by nemvar** [comment]
Fix this properly by creating a shared datum between all the projectiles and making a timer handle the time between message generation.

**Comment by yoyobatty** [comment]
Could add a cooldown to this instead, we did this on hippie if you wanna checkout that code.

**Comment by SkeletalElite** [comment]
> > I think getting rid of the buckshot lag is way more important than having a message when you get blasted with shotgun pellets.
> 
> that's what causes the lag, announce message is horribly optimized

thats exactly what Im saying. Way better off not having the message than freezing up when getting shot.

**Review by AnturK — CHANGES_REQUESTED** [review]
Yeah, this is pretty bad. Add per mob timeout indexed by projectile type ?

**Comment by yoyobatty** [comment]
https://github.com/HippieStation/HippieStation/blob/b4f2e2d83ec6343184c39812023c4e60d20c48fe/code/modules/projectiles/projectile.dm#L192

**Comment by carlarctg** [comment]
this is wip currently being rewokred

File: code/modules/projectiles/projectile.dm [hunk 2]
```diff
@@ -191,8 +193,11 @@
 			if(hitsound)
 				var/volume = vol_by_damage()
 				playsound(loc, hitsound, volume, TRUE, -1)
-			L.visible_message("<span class='danger'>[L] is hit by \a [src][organ_hit_text]!</span>", \
-					"<span class='userdanger'>You're hit by \a [src][organ_hit_text]!</span>", null, COMBAT_MESSAGE_RANGE)
+			if(!L.message_timeouts || L.message_timeouts <= world.time)
+				L.visible_message("<span class='danger'>[L] is hit by \a [src][organ_hit_text]!</span>", \
+						"<span class='userdanger'>You're hit by \a [src][organ_hit_text]!</span>", null, COMBAT_MESSAGE_RANGE)
+				if(lagshot)
```
Comment (AnturK) [inline code/modules/projectiles/projectile.dm]: No need for this variable, just check if timeout is set.

File: code/modules/projectiles/projectile.dm [hunk 3]
```diff
@@ -191,8 +193,11 @@
 			if(hitsound)
 				var/volume = vol_by_damage()
 				playsound(loc, hitsound, volume, TRUE, -1)
-			L.visible_message("<span class='danger'>[L] is hit by \a [src][organ_hit_text]!</span>", \
-					"<span class='userdanger'>You're hit by \a [src][organ_hit_text]!</span>", null, COMBAT_MESSAGE_RANGE)
+			if(!L.message_timeouts || L.message_timeouts <= world.time)
```
Comment (AnturK) [inline code/modules/projectiles/projectile.dm]: `if(!L.message_timeouts || !L.message_timeouts[type] || L.message_timeouts[type] <= world.time)`

**Review by AnturK — CHANGES_REQUESTED** [review]

**Comment by actioninja** [comment]
Stale

**Comment by carlarctg** [comment]
hello

File: code/modules/projectiles/projectile.dm (hunk 3 above)
Comment (carlarctg) [reply code/modules/projectiles/projectile.dm]: just (type)? nothing else?

Comment (AnturK) [reply code/modules/projectiles/projectile.dm]: That's what you set with that LAZYSET as key so yes.

**Comment by carlarctg** [comment]
i made your changes and it still isn't working @AnturK 
it doesn't stop the to_chat from happening, as far as i can tell in localhost testing with two emitters set to buckshot and min/max fire delay of 1

**Comment by AnturK** [comment]
I can't review code i don't see.

**Comment by carlarctg** [comment]
heck

**Comment by carlarctg** [comment]
working on this by the way, i'm an idiot so progress is slow but its happening

**Comment by actioninja** [comment]
In that case just open a new pr or ping someone to reopen this one when it's ready.

**Pull request closed without merging**